### DISCUSSION TITLE
Dynamic versioning with pyproject.toml

### DISCUSSION THREAD

**Discussion by rndubs** [pypa/setuptools#3630] [Q&A] [answered]
Hi Folks,

I'm using `setuptools` version 65.4.1 and python 3.7.2.

I am using a `pyproject.toml` file to package my module, and I would like to set the project version "dynamically" using a datetime-derived version number that is determined at build time. Currently when attempting to build the dist with `python3 -m build` (build version 0.8.0) I am getting an error, which is provided below.

I've read through the [dynamic metadata](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) section of the docs.

## Repo Layout

I'm currently attempting to achieve dynamic versioning as follows (only including relevant portions of the repo structure):

```
mymodule/
├── pyproject.toml
├── mymodule/
│       ├── __init__.py
```

The `mymodule.__init__.py` file:

```python
__version__ =  f"{1}.{2}.{3}"  # more complicated than this, but a simple example
```

A snippet from the `pyproject.toml` file:

```toml
[project]
name = "mymodule"
dynamic = ["version"]

[build-system]
requires = ["setuptools>=61.0.0", "setuptools-scm"]
build-backend = "setuptools.build_meta"

[tool.setuptools.packages.find]
where = ["."]
include = ["mymodule*"]

[tool.setuptools.dynamic]
version = {attr = "mymodule.__version__"}
```

## Current Behavior

Building a wheel and dist with `python3 -m build` currently results in the following error (not including the full traceback):
```
...  # full stacktrace snipped here
File "~/python-3.7.2/lib/python3.7/ast.py", line 55, in _convert_num
    raise ValueError('malformed node or string: ' + repr(node))
ValueError: malformed node or string: <_ast.Call object at 0x2aaab1027278>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/build-env-x7ix9t6m/lib/python3.7/site-packages/setuptools/config/expand.py", line 190, in read_attr
    return getattr(StaticModule(module_name, spec), attr_name)
  File "/build-env-x7ix9t6m/lib/python3.7/site-packages/setuptools/config/expand.py", line 86, in __getattr__
    raise AttributeError(f"{self.name} has no attribute {attr}") from e
AttributeError: mypackage has no attribute __version__
```

Configuring the `mymodule.__init__.py` file as follows results in a successful build, but without the desired dynamic versioning at build time:

```python
__version__ =  "1.2.3"
```

Is there a way to configure my project to include the dynamic versioning info by invoking a function?

**Comment by abravalheri** [comment, marked as answer]
Hi @rndubs, I cannot reproduce this error with the following experiment (in an isolated docker container):

```bash
> docker run --rm -it python:3.7.2 /bin/bash
python -m pip install -U pip 'setuptools==65.4.1' build
mkdir -p /tmp/myproj/mymodule
cd /tmp/myproj
echo -e '__version__ = f"{1}.{2}.{3}"' > mymodule/__init__.py
cat <<EOS > pyproject.toml
[project]
name = "mymodule"
dynamic = ["version"]

[build-system]
requires = ["setuptools>=61.0.0", "setuptools-scm"]
build-backend = "setuptools.build_meta"

[tool.setuptools.packages.find]
where = ["."]
include = ["mymodule*"]

[tool.setuptools.dynamic]
version = {attr = "mymodule.__version__"}
EOS
python -m build
```

This successfully creates the `mymodule-1.2.3-py3-none-any.whl` distribution. 

If I replace the `__init__.py` with the following

```bash
echo -e 'a = 4\nb = 5\nc = 6\n__version__ = f"{a}.{b}.{c}"' > mymodule/__init__.py
```

and run `python -m build` again, it will successfully create a `mymodule-4.5.6-py3-none-any.whl` distribution.

**Comment by agchitu** [comment]
This issue is still present in the current version of setuptools. 
It only works when `__version__` is a simple string and not the result of a process, not even formatting is allowed.
So `__version__ = "1.2.3"` works but `__version__ = f"{get_dynamic_version_returns_string()}"` will not work

**Reply by abravalheri** [reply]
> So __version__ = "1.2.3" works but __version__ = f"{get_dynamic_version_returns_string()}" will not work

@agchitu, please note that packages in the form of source code do not exist yet as an fully functional installed packages during build-time, so evaluating it is a complex business and execution errors/exceptions may happen (e.g. missing dependencies)[^2]. This is a natural limitation which occurs due to the nature of a build process.

Therefore calculating the version dynamically is generally error prone, and static definitions are recommended. The following tips may reduce the chances of error:

1. Make sure you don't use any custom package layout that needs to be re-worked during build-time (like remapping package directories)
2. Make sure that the module you are trying to execute does not import other dependencies or files that may not exist during build-time. The package does not exist yet and was not installed at build-time. So the majority of runtime behaviour is not available.
3. Make sure that the module you are trying to execute has all imports available as if it existed in a freshly produced blank virtual environment with only the dependencies in `pyproject.toml`'s `[buil-system] requires` table installed[^1]. Not even the current directory is added to `sys.path` by default. If you need extra entrances in `sys.path`, you need to add customisations via `setup.py`.
4. If you find any errors, please analyse the build log and look for error messages or warnings. Tools like `pip install` may hide the logs by default but you can use the verbosity options to reveal it (e.g. by adding `-vv`).

For debugging purposes, an interesting way to test if your module containing `__version__` is friendly to be executed during the build is to:

3a. Create a brand new empty virtual env and activate it
3b. `cd` into a unrelated directory without a folder with the same name of your package (e.g. `mkdir /tmp/staging && cd /tmp/staging`)
3c. Attempt to "execute" your module using the recipe `import_from_path` in https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly. Failures or exceptions when doing so may indicate your module cannot be suitable for extracting the value of `__version__` during build-time (and it may depend on runtime conditions that are not available prior to the package's installation).

If you still are finding trouble please open a issue in the tracker including a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) (please feel free to use my example above as a basis/template when creating your own reproducer). If opening a new ticket, please do include evidence that the module containing `__version__` can be executed in isolation as indicated in the steps 3a, 3b and 3c above.

[^1]: The majority of frontends and installers may use the `build isolation` feature defined in the PyPA standard.
[^2]: Build-time behaviour differs significantly from runtime; e.g., runtime dependencies may have not been installed yet.
